### DISCUSSION TITLE
Can't know if yii connection is closed

### DISCUSSION THREAD

**Discussion by Mitirrli** [yiisoft/yii2#19836] [General]
Hello everyone, I am now encountering a problem. 

There are many sleep processes in the online mysql database. After checking, I have roughly known the problem. Most of the problems are that some io operations are time-consuming and the connection is always occupied. What I want to do now is to monitor whether the database connection is closed within the specified time, and if it is not closed after the specified time, I can notify myself by email, so that I can catch some problems in time.

But I looked at the source code, and there is no corresponding event for closing the connection. I can't easily realize my idea, of course I can modify classMap, but this method is too rough.

Is there any way to replace the code without modifying the classMap? thanks~

**Comment by bizley** [comment]
IIRC there is no event sent when connection is closed since this is done outside the framework. Calling open() on the connection should be safe if it's open already but I'm not sure about that.

**Reply by Mitirrli** [reply]
Yes, it is safe, i did the test.

**Comment by schmunk42** [comment]
> Calling open() on the connection should be safe if it's open already but I'm not sure about that.

I recently tested this and from my tests you can basically close the connection at any time, it is reopened automatically if needed.

@Mitirrli 
Do you have long running CLI tasks (queue, cron, ...) which might open the connection early?
Do you use persistent connections?
Which db?

**Reply by Mitirrli** [reply]
Q: Do you have long running CLI tasks (queue, cron, ...) which might open the connection early?  
A: yes

Q: Do you use persistent connections?  
A: no

Q: Which db?
A: mysql

---------

Yes, you are right.

I now have an export tasks, which is very time-consuming. Of course, it’s not just about export tasks, I also need to request third-party interfaces, which is also time-consuming.

I close the database connection before the operation. After operation, if I query the database, it will automatically reconnect. Framework is correct, i can use `Yii::$app->db->close();`   to close connection.

At present, I want to monitor the connection closing time. For example, if it exceeds 8 seconds, I can send an email to myself. In this way, I can deal with it in time to prevent the number of connections from being heavily occupied.

I can't think of a better way, except for the following use of ClassMap to override Connection.

```php
# Connection.php
protected function initConnection()  
{  
    ...

    $this->trigger(self::EVENT_AFTER_OPEN);  
    $this->connectInitTime = time();  
}

public function close()  
{  
    if ($this->isActive) {
        if ((time() - $this->connectInitTime) >= 8) {
	    // do sth to notify
	}
    }  

    ...
}
```

**Reply by schmunk42** [reply]
To trigger an event would be fine IMHO, could you come up with a PR for that.

Also the rest sounds reasonable to me.

**Reply by Mitirrli** [reply]
Ok😃, I'll do it when I'm free.
